src/core/utils/logger.cc

```diff
@@ -0,0 +1,273 @@
+// Python Logger class
+//------------------------------------------------------------------------------
+namespace py {
+
+
+class DefaultLogger : public XObject<DefaultLogger>
+{
+  private:
+    std::unique_ptr<std::string> prefix_;
+    bool use_colors_;
+    size_t : 56;
+
+    void m__init__(const PKArgs&) {}
+
+    void m__dealloc__() { prefix_ = nullptr; }
+
+    void debug(const PKArgs& args) {
+      auto msg = args[0].to_string();
+      print_message(msg, *prefix_, use_colors_);
+    }
+    // oobj warning();
+
+  public:
+    static oobj make(const dt::log::Logger& logger) {
+      robj rtype(reinterpret_cast<PyObject*>(&DefaultLogger::type));
+      oobj resobj = rtype.call();
+      DefaultLogger* dlogger = DefaultLogger::cast_from(resobj);
+      dlogger->prefix_ = std::make_unique<std::string>(logger.prefix_);
+      dlogger->use_colors_ = logger.use_colors_;
+      return resobj;
+    }
+
+    static void impl_init_type(XTypeMaker& xt) {
+      xt.set_class_name("DefaultLogger");
+      static PKArgs args_init(0, 0, 0, false, false, {}, "__init__", nullptr);
+      static PKArgs args_debug(1, 0, 0, false, false, {"msg"}, "debug", nullptr);
+      xt.add(CONSTRUCTOR(&DefaultLogger::m__init__, args_init));
+      xt.add(DESTRUCTOR(&DefaultLogger::m__dealloc__));
+      xt.add(METHOD(&DefaultLogger::debug, args_debug));
+    }
+
+    static void init_once() {
+      static bool initalized = false;
+      if (initalized) return;
+      DefaultLogger::init_type(nullptr);
+      initalized = true;
+    }
+};
+
+
+
+
+}  // namespace py
+namespace dt {
+namespace log {
+
+
+
+//------------------------------------------------------------------------------
+// Section
+//------------------------------------------------------------------------------
+
+Section::Section(Logger* logger)
+  : logger_(logger) {}
+
+
+Section::~Section() {
+  logger_->end_section();
+}
+
+
+
+
+//------------------------------------------------------------------------------
+// Message
+//------------------------------------------------------------------------------
+
+Message::Message(Logger* logger, bool warn)
+  : logger_(logger), emit_as_warning_(warn) {}
+
+
+Message::~Message() {
+  logger_->emit(std::move(out_).str(), emit_as_warning_);
+}
+
+
+
+ff::ff(int w, int p, double v)
+  : width(w), precision(p), value(v) {}
+
+template <>
+Message& Message::operator<<(const ff& f) {
+  out_ << std::fixed << std::setw(f.width)
+       << std::setprecision(f.precision)
+       << f.value;
+  return *this;
+}
+
+
+plural::plural(size_t n, const std::string& s)
+  : count_(n), str_(s.c_str()) {}
+
+plural::plural(size_t n, const char* s)
+  : count_(n), str_(s) {}
+
+template <>
+Message& Message::operator<<(const plural& pl) {
+  out_ << pl.count_ << ' ';
+  out_ << pl.str_;
+  if (pl.count_ != 1) {  // see misc.py for more advanced pluralization rules
+    out_ << 's';
+  }
+  return *this;
+}
```

Comment:
May be one day we will port the advanced pluralization code from Python to C++, and then expose it back to Python as well as use it from C++.

Reply:
Yeah, I think we can defer that to that time when the more advanced rules are actually needed. Until now we've been mostly pluralizing words like `row`, `column`, `thread` or `file`.